### ISSUE TITLE
dotnet tool update does not work when only pre-releases are available

### ISSUE THREAD

**Issue by adamralph** [dotnet/cli#10382]
## Steps to reproduce

(At the time of writing.)

- `dotnet tool install -g --version 1.0.0-alpha.15 minver-cli`
- `dotnet tool update -g minver-cli`

## Expected  behavior

The package is updated to `1.0.0-alpha.16` (or a later pre-release).

## Actual behavior

```
error NU1103: Unable to find a stable package minver-cli with version
error NU1103:   - Found 2 version(s) in NuGet [ Nearest version: 1.0.0-alpha.16 ]
error NU1103:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
error NU1103:   - Found 0 version(s) in C:\Program Files\dotnet\sdk\NuGetFallbackFolder
Tool 'minver-cli' failed to update due to the following:
The tool package could not be restored.
Tool 'minver-cli' failed to install. This failure may have been caused by:

* You are attempting to install a preview release and did not use the --version option to specify the version.
* A package by this name was found, but it was not a .NET Core tool.
* The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.
* You mistyped the name of the tool.
```

But `dotnet tool update` does not have a `--version` option!

Also, the first line in that output looks wrong.

## Environment data
`dotnet --info` output:

```
.NET Core SDK (reflecting any global.json):
 Version:   2.1.500
 Commit:    b68b931422

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.500\

Host (useful for support):
  Version: 2.1.6
  Commit:  3f4f8eebd8

.NET Core SDKs installed:
  1.0.4 [C:\Program Files\dotnet\sdk]
  1.1.0 [C:\Program Files\dotnet\sdk]
  2.0.0 [C:\Program Files\dotnet\sdk]
  2.0.2 [C:\Program Files\dotnet\sdk]
  2.0.3 [C:\Program Files\dotnet\sdk]
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.301 [C:\Program Files\dotnet\sdk]
  2.1.302 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by livarcocc** [comment]
@wli3

**Comment by wli3** [comment]
This issue root back to https://github.com/nuget/home/issues/912

**Comment by jonstelly** [comment]
@wli3 that issue seems to be about floating version numbers, but the update command is simply missing a version argument.

Looking at the CLI code (assuming I'm reading things right):

Install
[ToolInstallGlobalOrToolPathCommand.cs option](https://github.com/dotnet/cli/blob/master/src/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs#L58)
[ToolInstallGlobalOrToolPathCommand.cs install](https://github.com/dotnet/cli/blob/master/src/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs#L126)

Update
[ToolUpdateCommand.cs install](https://github.com/dotnet/cli/blob/master/src/dotnet/commands/dotnet-tool/update/ToolUpdateCommand.cs#L111)

Seems like update doesn't accept the version parameter and doesn't specify a version when installing, but since update does an uninstall then an install, it seems like just adding the same support for a version parameter should be pretty straight-forward.

**Comment by wli3** [comment]
@jonstelly 

1. should _update_ accept a version? In UX perspective i think it will cause more confusion. What happens when the existing version is higher than the older version? and in many case, i believe the user intent is "to this version, update or downgrade". so maybe instead of version on _update_, it should be a option in _install_.

2. will above solution solve the issue? I believe is not. The issue is intent to update to the latest including prerelease version, not a specific version. And today there is such operation in nuget.xplat (while in windows only nuget.exe there is a -pre flag). That's why we are blocked by NuGet/Home#912

**Comment by jonstelly** [comment]
Fair points, but in my opinion that linked nuget issue seems unlikely to be fixed anytime soon (the captured requirements don't clearly spell out expected behavior in different scenarios), and after that issue gets resolved, getting the global tool install to support that sort of version specification seems like a fair chunk of work.

I do see some more discussion about tool installs/updates for prerelease specifically here: https://github.com/dotnet/sdk/issues/2551

I'll admit this isn't a terrible problem (probably not a lot of people affected by this and the workaround is clear enough), but it would be nice to have a better option in the short-term.

Maybe a change to the install command's behavior:

`dotnet tool install -g mytool` - does nothing and warns if mytool is already installed, the user isn't specifying a version, just asking to make sure that the tool is installed
`dotnet tool install -g --version 1.2.3.4 mytool` - does nothing and warns if a newer or equal version of mytool is already installed, or succeeds if an older version of mytool is already installed (by uninstalling the old and installing the new)
